tests/helpers/mapbox-gl-stub.js

```diff
@@ -0,0 +1,213 @@
+    getStyle: () => ({ sources: {}, layers: {} }),
+    getZoom: () => {},
+    off: () => {},
+    on: () => {},
+    queryRenderedFeatures: () => [],
+    querySourceFeatures: () => [],
+    removeControl: () => {},
+    removeLayer: () => {},
+    resize: () => {},
+    setFilter: () => {},
+    setPaintProperty: () => {},
+    unproject: () => ({ lat: 0, lng: 0, toArray: () => [0, 0] }),
+
+    scrollZoom: UserInteractionHandlerStub,
+    boxZoom: UserInteractionHandlerStub,
+    dragRotate: UserInteractionHandlerStub,
+    dragPan: UserInteractionHandlerStub,
+    keyboard: UserInteractionHandlerStub,
+    doubleClickZoom: UserInteractionHandlerStub,
+    touchZoomRotate: UserInteractionHandlerStub,
+  },
+  draw: {
+    add: () => {},
+    set: () => {},
+    getAll: () => ({ features: [] }),
+    getSelected: () => [],
+    getSelectedPoints: () => ({ features: [] }),
+    getSelectedIds: () => [],
+    getMode: () => 'simple_select',
+    changeMode: () => {},
+  },
+};
+
+// define a new mapbox-gl class with the layout also stubbed in.
+// this will replace the ember-mapbox-gl bindings during tests
+@classNames('mapbox-gl')
+export class MapboxGl extends Component {
+  mapLoaded = () => {}
+
+  // current test context, used to contain reference to
+  // test scope for dependency injections & stubbing
+  // mapbox-gl methods
+  mapboxEventStub = defaultMapboxEventStub;
+
+  init(...args) {
+    // mapLoaded is an ember-mapbox-gl convenience
+    // which specifically sends out the "target"
+    // as the map instance
+    this.map = this.mapboxEventStub.mapInstance;
+    this.mapLoaded(this.map);
+
+    super.init(...args);
+  }
+
+  // this is a layout stub which is necessary to replace contextual components
+  // that descend from {{mapbox-gl}}
+  // as more ember-mapbox-gl contextual components are needed for stubbing,
+  // they must be added here
+  layout = hbs`
+    {{!-- Highlighted Layer Handling --}}
+
+    {{yield (assign 
+      (hash
+        on=(component 'mapbox-gl-on'
+          eventSource=this.mapboxEventStub.mapInstance
+        )
+        source=(component 'mapbox-gl-source'
+          map=this.map
+        )
+        layer=(component 'mapbox-gl-layer'
+          map=this.map
+        )
+
+        draw=this.mapboxEventStub.draw
+
+        mapInstance=this.mapboxEventStub.mapInstance
+      )
+      this.mapboxEventStub
+    )}}
+  `;
+}
+
+// stub for mapbox-gl-on, the ember-mapbox-gl interface for binding events
+@tagName('')
+export class MapboxOn extends Component {
+  // this component gets "positional params" so that it can be invoked as follows:
+  // {{map.on 'someEvent' (action 'myAction')}}
+  static positionalParams = ['event', 'optionalLayerId', 'action'];
+
+  // the core functionality of mapbox-gl-on is handled on init by binding
+  // the passed event name with the passed action.
+  // bindOnInvocation is used to pull in test context the moment the event itself is
```

Comment:
bindOnInvocation is a method where exactly? on what class? Feels that context is necessary here for those of us who are not as fully immersed in this world as you

Reply:
oops it should not be there anymore!